src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/Frame.FeatureCollection.cs

```diff
@@ -235,6 +235,18 @@ void IHttpResponseFeature.OnCompleted(Func<object, Task> callback, object state)
                 throw new InvalidOperationException(CoreStrings.CannotUpgradeNonUpgradableRequest);
             }
 
+            if (_wasUpgraded)
+            {
+                throw new InvalidOperationException(CoreStrings.UpgradeCannotBeCalledMultipleTimes);
+            }
+
+            if (!ServiceContext.Resources.UpgradedConnections.TryLockOne())
+            {
+                throw new InvalidOperationException(CoreStrings.UpgradedConnectionLimitReached);
+            }
+
+            _wasUpgraded = true;
```

Comment:
@natemcmaster Do you want to change this to call NormalConnections.ReleaseOne() here (and of course avoid calling it twice), or do you want to remove MaxConcurrentUpgradedConnections? I'll approve either way, but we shouldn't have a separate upgraded connection limit that still applies to the normal connection limit.